hkmc2/shared/src/main/scala/hkmc2/semantics/ucs/Normalization.scala

```diff
@@ -105,78 +115,288 @@ class Normalization(elaborator: Elaborator)(using raise: Raise, ctx: Ctx):
+            else // If any errors were raised, we skip the branch.
+              log("BROKEN"); normalizeImpl(alternative)
+          case S(S(pat: PatternSymbol)) => mode match
+            // Note: `argsOpt` is supposed to be used in following cases, but
+            // the current implementation does not use it. The future version
+            // should properly handle the pattern arguments.
+            case MatchMode.Default =>
+              normalizeExtractorPattern(scrutinee, pat, ctor, consequent, alternative)
+            case MatchMode.StringPrefix(prefix, postfix) =>
+              normalizeStringPrefixPattern(scrutinee, pat, ctor, postfix, consequent, alternative)
+            case MatchMode.Annotated(annotation) => annotation.symbol.flatMap(_.asObj) match
+              case S(symbol) if symbol === ctx.builtins.compile =>
+                normalizeCompiledPattern(scrutinee, pat, ctor, argsOpt, mode, consequent, alternative)
+              case S(_) =>
+                warn(msg"Unknown annotation on pattern" -> annotation.toLoc,
+                  msg"Note: only `@compile` is supported." -> N)
+                normalizeExtractorPattern(scrutinee, pat, ctor, consequent, alternative)
+              case N =>
+                // Name resolution should have already reported an error. We
+                // treat this as an extractor pattern.
+                normalizeExtractorPattern(scrutinee, pat, ctor, consequent, alternative)
+    case Split.Let(v, _, tail) if vs has v =>
+      log(s"LET: SKIP already declared scrutinee $v")
+      normalizeImpl(tail)
+    case Split.Let(v, rhs, tail) =>
+      log(s"LET: $v")
+      Split.Let(v, rhs, normalizeImpl(tail)(using vs + v))
+    case Split.Else(default) =>
+      log(s"DFLT: ${default.showDbg}")
+      Split.Else(default)
+    case Split.End => Split.End
+  
+  /** Check whether the number of parameters in class-like patterns matches the
+   *  number in their definition, and whether each parameter is accessible.
+   */
+  private def validateClassPattern(
+      ctorTerm: Term,
+      ctorSymbol: ClassSymbol,
+      argsOpt: Opt[Ls[Pattern.Argument]]
+  ): Bool =
+    // Obtain the `classHead` used for error reporting and the parameter list
+    // from the class definitions.
+    val (classHead, paramsOpt) =
+      // Unfortunately, class symbols of `MatchResult` and `MatchFailure`
+      // are forged in the elaborator state, so it does not have `defn`.
+      // TODO(ucs): do not make special case for them
+      if ctorSymbol == State.matchResultClsSymbol then
+        (ctorSymbol.id: Located) -> matchResultClassParamOpt
+      else if ctorSymbol == State.matchFailureClsSymbol then
+        (ctorSymbol.id: Located) -> matchFailureClassParamOpt
+      else
+        ctorSymbol.defn match
+          case N => lastWords(s"Class ${ctorSymbol.name} does not have a definition")
+          case S(cd) => ctorSymbol.id -> cd.paramsOpt
```

Comment:
I’d like to know if there’s a solution to this.

Reply:
Can we get symbols for `MatchResult` and `MatchFailure` by adding them to `Prelude.mls` and then using `assumeBuiltinCls`?

Reply:
The simplest solution is to just put the definitiosn there yourself. Any problem with that approach? Otherwise, we'd have to declare these classes in the `Prelude` and special-case their lowering, but that seems a bit troublesome.